qcodes/tests/dataset/test_data_set_cache.py

```diff
@@ -331,6 +333,150 @@ def test_cache_complex_array_param_in_1d(experiment, DAC, channel_array_instrume
+
+    if cache_too_small:
+        meas.set_shapes(detect_shape_of_measurement(
+            meas_parameters,
+            (int(ceil(n_points_outer/2)), n_points_inner))
+        )
+    else:
+        meas.set_shapes(detect_shape_of_measurement(
+            meas_parameters,
+            (n_points_outer, n_points_inner))
+        )
+
+    expected_shapes = {
+        'dummy_dmm_v1': (n_points_outer, n_points_inner),
+        'dummy_channel_inst_ChanA_multi_setpoint_param_this': (n_points_outer, n_points_inner, 5),
+        'dummy_channel_inst_ChanA_multi_setpoint_param_that': (n_points_outer, n_points_inner, 5),
+        'dummy_channel_inst_ChanA_thisparam': (n_points_outer, n_points_inner),
+        'dummy_channel_inst_ChanA_thatparam': (n_points_outer, n_points_inner),
+        'dummy_channel_inst_ChanA_this': (n_points_outer, n_points_inner, 5, 3),
+        'dummy_channel_inst_ChanA_that': (n_points_outer, n_points_inner, 5, 3),
+        'dummy_channel_inst_ChanA_this_5_3': (n_points_outer, n_points_inner, 5, 3),
+        'dummy_channel_inst_ChanA_this_2_7': (n_points_outer, n_points_inner, 2, 7),
+        'dummy_channel_inst_ChanA_dummy_array_parameter': (n_points_outer, n_points_inner, 5),
+        'dummy_channel_inst_ChanA_dummy_complex_array_parameter': (n_points_outer, n_points_inner, 5),
+        'dummy_channel_inst_ChanA_dummy_complex': (n_points_outer, n_points_inner),
+        'dummy_channel_inst_ChanA_dummy_parameter_with_setpoints': (n_points_outer, n_points_inner, pws_n_points),
+        'dummy_channel_inst_ChanA_dummy_parameter_with_setpoints_complex': (n_points_outer, n_points_inner, pws_n_points)
+    }
+
+    if not cache_too_small:
+        assert meas._shapes == expected_shapes
+
+    with meas.run(write_in_background=bg_writing) as datasaver:
+        dataset = datasaver.dataset
+        _assert_parameter_data_is_identical(dataset.get_parameter_data(), dataset.cache.data())
+        n_points_measured = 0
+        for v1 in np.linspace(-1, 1, n_points_outer):
+            for v2 in np.linspace(-1, 1, n_points_inner):
+                n_points_measured += 1
+                DAC.ch1.set(v1)
+                DAC.ch2.set(v2)
+                meas_vals = [(param, param.get()) for param in meas_parameters]
+
+                datasaver.add_result((DAC.ch1, v1),
+                                     (DAC.ch2, v2),
+                                     *meas_vals)
+                datasaver.flush_data_to_database(block=True)
+                param_data_trees = dataset.get_parameter_data()
+                cache_data_trees = dataset.cache.data()
+
+                _assert_partial_cache_is_as_expected(
+                    cache_data_trees,
+                    expected_shapes,
+                    n_points_measured,
+                    param_data_trees,
+                    cache_too_small
+                )
+    cache_data_trees = dataset.cache.data()
+    param_data_trees = dataset.get_parameter_data()
+
```

Comment:
are they not identical when the measurement is complete?
```suggestion
    _assert_parameter_data_is_identical(param_data_trees, cache_data_trees)
```